README.md
# flutter_celik_api

A Flutter plugin for reading Serbian ID Cards. Works on Windows/Linux/MacOS.
Written from scratch (APDU Commands).

Connection to card-reader is done through `flutter_pcsc` but you may use your own custom provider.

## Usage

### Pre-requisite

`flutter_pcsc` needs:

- A PCSC smartcard reader.

- On linux, `pcscd` & `libpcsclite1` needs to be installed.

- For macOS application to be able to use smartcard, the following entitlement should be set: `com.apple.security.smartcard` (in DebugProfile.entitlements & Release.entitlements files).  
  If not set correctly, the context won't be able to be established.

### Example

```dart

CelikAPIAllData result = await CelikAPI().readAllData();
print(result.data);
print(result.image); // base64

```

### Features

- [x] Reading data  
- [x] Reading image  
- [ ] Cert installation  
- [ ] Verification  
- [ ] Signing data  
- [ ] Changing card pin


CHANGELOG.md
# Change Log

### version: 1.0.1+2

- Better, more descriptive error handling
- String error messages instead of `Exception`s

### version: 1.0.1+1

- Minor comments fixes
- Exporting CelikApiAllData class

### version: 1.0.1

- Added `CelikApi.disconnectCard()`
- Added posiibility for custom implementation, using `CelikDataAPI` and `CelikAPDUBase`
- Better file structure
- Improved comments
- Fixed library name
- Fix: now running PCSC in `newIsolate = true`
- Simple test added
- Updated `README.md`

### version: 1.0.0+1

Test version


test/flutter_celik_api_test.dart
import 'package:flutter_celik_api/flutter_celik_api.dart';
import 'package:test/test.dart';

const pin = const String.fromEnvironment("pin");

void main() {
  test("CelikAPI - readAllData()", () async {
    final response = await CelikAPI().readAllData();

    expect(response.data, isNot(equals({})));
    expect(response.image, isNot(equals("")));
    expect(response.data[CelikTag.iss], equals("SRB"));
  });

  group("CelikAPI - Single tests", () {
    final celik = CelikAPI();
    setUpAll(() async {
      await celik.start();
      final readers = await celik.getReaders();
      if (readers.isEmpty) throw "No readers available.";
      await celik.connectCard(readers[0]);
    });
    tearDownAll(() async {
      await celik.end();
    });

    /*
    test("- sign & verify", () async {
      if (pin == "")
        throw "No pin specified. Specify with --dart-define=\"pin=[PIN]\" argument.";

      const testStr = "Hello World!";
      final signature = await celik.signData(
        utf8.encode(testStr),
        pin,
      );
      final isOk = await celik.verifySignature(
        signature,
        utf8.encode(testStr),
      );

      expect(isOk, isTrue);
    }); 
    */
  });
}


lib/src/apdu.dart
class APDUStandard {
  static const List<int> successCode = [0x90, 0];
  static const int maxReadBytes = 32;
}

class APDUCommands {
  static List<int> initAID(List<int> aid) => [
        ...[0x00, 0xA4, 0x04, 0x00],
        aid.length,
        ...aid
      ];

  static List<int> selectFile(List<int> name) => [
        ...[0x00, 0xA4, 0x08, 0x00],
        name.length,
        ...name
      ];

  static List<int> readByteData(int bytesOffset, int bytesLength) => [
        ...[0x00, 0xB0],
        bytesOffset >> 8, // or 0
        bytesOffset & 0xFF,
        bytesLength
      ];

  static List<int> verify(List<int> pin) => [
        ...[0x00, 0x20, 0x00, 0x01],
        pin.length,
        ...pin,
        0x00
      ];
}

class APDUResponse {
  final List<int> _raw;
  late List<int> _statusCode;

  APDUResponse(this._raw) {
    _statusCode = _raw.getRange(_raw.length - 2, _raw.length).toList();
  }

  bool isOk() {
    return _statusCode[0] == APDUStandard.successCode[0] &&
        _statusCode[1] == APDUStandard.successCode[1];
  }

  List<int> data() => _raw.getRange(0, _raw.length - 2).toList();

  get statusCode => _statusCode;
}


lib/src/celik_api.dart
import 'dart:convert';

import './apdu.dart';
import './celik_apdu_base.dart';
import './constants.dart';
import './util.dart';

export './celik_apdu_base.dart' show CelikAPDUBase;

/// Class that defines the methods for reading data
/// Extend and override createCelikAPDU()
abstract class CelikDataAPI {
  /// Should be overriden with your custom PCSC logic
  CelikAPDUBase createCelikAPDU();

  /// Read binary data from CelikFile
  Future<List<int>> readBinaryData(CelikFile file) async {
    CelikAPDUBase celik = createCelikAPDU();
    APDUResponse initResponse = await celik.init();
    if (!initResponse.isOk()) throw Exception("Init failed.");

    APDUResponse selectFileResponse = await celik.selectFile(file);
    if (!selectFileResponse.isOk()) throw Exception("Selecting file failed.");

    APDUResponse dataResponse = await celik.readFile();
    if (!dataResponse.isOk()) throw Exception("Reading data error!");

    return dataResponse.data();
  }

  /// Verify PIN
  Future<void> verify(List<int> pin) async {
    CelikAPDUBase celik = createCelikAPDU();
    APDUResponse initResponse = await celik.init();
    if (!initResponse.isOk()) throw Exception("Init failed.");

    APDUResponse verifyResponse = await celik.verifyPIN(pin);
    if (!verifyResponse.isOk()) {
      if (verifyResponse.statusCode[0] == 0x63) {
        final sw2 = verifyResponse.statusCode[1];
        if (sw2 >= 0xC0 && sw2 <= 0xCF)
          throw "Pin incorrect. ${sw2 - 0xC0} tries left.";
      }
      throw Exception("Pin not correct.");
    }
  }

  /// Read data (structured) from CelikFile
  Future<Map<CelikTag, String>> readData(CelikFile file) async {
    return transformData(await readBinaryData(file));
  }

  /// Reads image data
  Future<String> readImageData() async {
    CelikAPDUBase celik = createCelikAPDU();
    APDUResponse initResponse = await celik.init();
    if (!initResponse.isOk()) throw Exception("Init failed.");

    APDUResponse selectFileResponse =
        await celik.selectFile(CelikFile.photoFile);
    if (!selectFileResponse.isOk()) throw Exception("Selecting file failed.");

    APDUResponse dataResponse = await celik.readFile(startOffset: 8);
    if (!dataResponse.isOk()) throw Exception("Reading image data error!");

    return base64.encode(dataResponse.data());
  }

  /// Returns all data on a card in a single call
  Future<CelikAPIAllData> readAllData() async => CelikAPIAllData({
        ...await readData(CelikFile.documentFile),
        ...await readData(CelikFile.residenceFile),
        ...await readData(CelikFile.personalFile),
      }, await readImageData());
}

class CelikAPIAllData {
  final Map<CelikTag, String> data;
  final String image;
  CelikAPIAllData(this.data, this.image);
}


lib/src/constants.dart
class CelikConstants {
  static const List<int> lkAID = [
    0xF3,
    0x81,
    0x00,
    0x00,
    0x02,
    0x53,
    0x45,
    0x52,
    0x49,
    0x44,
    0x01
  ];

  static List<int> getFile(CelikFile f) {
    switch (f) {
      case CelikFile.documentFile:
        return [0x0F, 0x02];
      case CelikFile.personalFile:
        return [0x0F, 0x03];
      case CelikFile.residenceFile:
        return [0x0F, 0x04];
      case CelikFile.photoFile:
        return [0x0F, 0x06];
      case CelikFile.authCertificate:
        return [0x0F, 0x08];
      case CelikFile.qualifiedCertificate:
        return [0x0F, 0x10];
      case CelikFile.intermCertificate:
        return [0x0F, 0x11];
      case CelikFile.encryptedPrivateKey:
        return [0x0F, 0x09];
      case CelikFile.encryptedPinAndSecret:
        return [0x0F, 0x13];
      case CelikFile.encryptionXOR:
        return [0x0F, 0xA1];
      case CelikFile.firstUnknown:
        return [0x0F, 0xA3];
      case CelikFile.test:
        return [0x0F, 0x12];
    }
  }

  static const Map<CelikTag, String> tags = {
    CelikTag.error: "ERR!",
    CelikTag.iss: "ISS",
    CelikTag.docRegNo: "DOC_REG_NO",
    CelikTag.id: "ID",
    CelikTag.idDocRegNo: "ID<DOC_REG_NO>",
    CelikTag.issDate: "ISSUING_DATE",
    CelikTag.expDate: "EXPIRY_DATE",
    CelikTag.issAuthority: "ISSUING_AUTHORITY",
    CelikTag.sc1: "SC",
    CelikTag.sc2: "SC",
  };

  static CelikTag tagFromInt(int tag) {
    int index = tag - _tagIndexStart;
    if (index >= CelikTag.values.length) {
      return CelikTag.error;
    }
    return CelikTag.values[index];
  }

  static String tagToString(CelikTag tag) {
    return tags[tag] ?? "-";
  }
}

enum CelikFile {
  documentFile,
  personalFile,
  residenceFile,
  photoFile,
  qualifiedCertificate,
  authCertificate,
  firstUnknown,
  encryptedPrivateKey,
  encryptedPinAndSecret,
  encryptionXOR,
  intermCertificate,
  test
}

const _tagIndexStart = 1545 - 1; // 1543 = error

extension CelikTagId on CelikTag {
  int get id => index + _tagIndexStart;
}

enum CelikTag {
  error, // = 1543
  iss,
  docRegNo,
  id,
  idDocRegNo,
  issDate,
  expDate,
  issAuthority,
  sc1,
  sc2,
  n1,
  n2,
  n3,
  n4,
  personalNumber,
  surname,
  givenName,
  parentGivenName,
  sex,
  placeOfBirth,
  communityOfBirth,
  stateOfBirth,
  dateOfBirth,
  countryOfBirth,
  state,
  community,
  place,
  street,
  houseNumber,
  houseLetter,
  entrance,
  floor,
  n5,
  n6,
  apartmentNumber,
  n7,
  addressDate,
  addressType,
}


lib/src/celik_crypto.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:aes_crypt_null_safe/aes_crypt_null_safe.dart';
import 'package:asn1lib/asn1lib.dart';
import 'package:crypto/crypto.dart';
import 'package:x509/x509.dart';

import './constants.dart';
import './celik_api.dart';

final signingAlgorithm = algorithms.signing.rsa.sha256;

mixin CelikCrypto on CelikDataAPI {
  Future<bool> verifySignature(Uint8List signature, List<int> data) async {
    final publicKey = await _getPublicKey();

    return publicKey.createVerifier(signingAlgorithm).verify(
          Uint8List.fromList(data),
          Signature(signature),
        );
  }

  Future<Uint8List> signData(List<int> data, String password) async {
    final docData = await readData(CelikFile.documentFile);
    final id = docData[CelikTag.docRegNo]!;
    final xorKey = utf8.encode("ID$id\u0001$password");
    final xorKeyHash = sha1.convert(xorKey).bytes;

    final encryptedData = await readBinaryData(CelikFile.encryptedPinAndSecret);
    final encryptedPIN = encryptedData.sublist(4, 4 + 8);
    final encryptedSecretKey = encryptedData.sublist(17, 17 + 32);

    final xorFile = await readBinaryData(CelikFile.encryptionXOR);

    final pin = _xorDecryptPIN(
      encryptedPIN,
      xorKeyHash,
      xorFile,
    );
    final secretKey = _xorDecryptSecretKey(
      encryptedSecretKey,
      xorKeyHash,
      xorFile,
    );

    await verify(pin);

    final publicKey = await _getPublicKey();

    final privateKey = _reconstructPrivateKey(
      await _getDecryptedPrivateKeyASN1(secretKey),
      publicKey.modulus,
      publicKey.exponent,
    );

    return privateKey.createSigner(signingAlgorithm).sign(data).data;
  }

  Future<RsaPublicKey> _getPublicKey() async {
    final standardCertificate =
        (await readBinaryData(CelikFile.authCertificate)).sublist(4);

    final pemIterable = parsePem(utf8.decode(standardCertificate));
    return pemIterable.firstWhere((x) => x is RsaPublicKey) as RsaPublicKey;
  }

  Future<Uint8List> _getDecryptedPrivateKeyASN1(List<int> secretKey) async {
    final encryptedPrivateKey =
        await readBinaryData(CelikFile.encryptedPrivateKey);
    final sha1hash = sha1.convert(secretKey).bytes;
    final iv = sha1hash.sublist(0, 16);

    final crypt = AesCrypt();
    crypt.aesSetParams(
      Uint8List.fromList(secretKey),
      Uint8List.fromList(iv),
      AesMode.cfb,
    );

    final decrypted = crypt.aesDecrypt(Uint8List.fromList(encryptedPrivateKey));
    return decrypted.sublist(4);
  }
}

RsaPrivateKey _reconstructPrivateKey(
  Uint8List privateKeyASN1,
  BigInt modulus,
  BigInt publicExponent,
) {
  final parser = ASN1Parser(privateKeyASN1);
  List<BigInt> privateKeyParts = [];
  while (parser.hasNext()) {
    privateKeyParts.add(BigInt.parse(
      parser.nextObject().toHexString(),
    ));
  }
  final prime1 = privateKeyParts[0];
  final prime2 = privateKeyParts[1];
  // final primeExponent1 = privateKeyParts[2];
  // final primeExponent2 = privateKeyParts[3];
  // final crtCoeficient = privateKeyParts[4];

  final fi = (prime1 - BigInt.one) * (prime2 - BigInt.one);
  final privateExponent = publicExponent.modInverse(fi);

  return RsaPrivateKey(
    privateExponent: privateExponent,
    firstPrimeFactor: prime1,
    secondPrimeFactor: prime2,
    modulus: modulus,
  );
}

List<int> _xorDecryptPIN(
  List<int> pin,
  List<int> xorKeyHash,
  List<int> xorValueWhole,
) {
  List<int> xorValue = xorValueWhole.sublist(4, 4 + 8);

  List<int> xoredValue = [];
  for (var i = 0; i < 8; i++) {
    xoredValue.add(xorValue[i] ^ pin[i] ^ xorKeyHash[i]);
  }
  return xoredValue;
}

List<int> _xorDecryptSecretKey(
  List<int> secretKey,
  List<int> xorKeyHash,
  List<int> xorValueWhole,
) {
  List<int> xorValue = xorValueWhole.sublist(4, 4 + 16);

  List<int> xoredValue = [];
  for (var i = 0; i < 32; i++) {
    xoredValue.add(xorValue[i % 16] ^ secretKey[i] ^ xorKeyHash[i % 16]);
  }
  return xoredValue;
}


lib/flutter_celik_api.dart
library flutter_celik_api;

import 'package:flutter/foundation.dart' show kDebugMode;
import 'package:flutter_pcsc/flutter_pcsc.dart';

import './src/constants.dart';
import './src/celik_api.dart';
import './src/celik_crypto.dart';
import './src/util.dart' show transformPCSCError;

export './src/constants.dart' show CelikFile, CelikTag;
export './src/celik_api.dart' show CelikAPDUBase, CelikDataAPI, CelikAPIAllData;

/// CelikAPDUBase, based on flutter_pcsc
class CelikAPDU extends CelikAPDUBase {
  final CardStruct card;
  CelikAPDU(this.card) : super();

  @override
  Future<List<int>> transmit(List<int> bytes) async {
    try {
      return await Pcsc.transmit(card, bytes, newIsolate: true);
    } catch (error) {
      throw transformPCSCError(error.toString());
    }
  }
}

/// Main user-friendly class.
/// * Implementation of CelikDataAPI.
/// * Communication with card-reader is done through flutter_pcsc package.
///
/// Simple usage: ```
/// await CelikAPI().readAllData()
/// ```
class CelikAPI extends CelikDataAPI with CelikCrypto {
  /// flutter_pcsc Context
  int? ctx;

  /// flutter_pcsc Card object
  CardStruct? card;

  /// Creates CelikAPDUBase based on `flutter_pcsc` package
  @override
  CelikAPDUBase createCelikAPDU() => CelikAPDU(card!);

  /// Check if context is initiated - or throw!
  checkInit() {
    if (ctx == null) {
      throw "No context detected. Please run start()";
    }
  }

  /// Check if card is present - or throw!
  checkCard() {
    if (card == null) {
      throw "No card detected! Init a card.";
    }
  }

  /// Called to initialize the PCSC Context. Initiates the connection to a card
  ///
  /// * Must be called before any other methods
  Future<void> start() async {
    await end();

    try {
      int ctx = await Pcsc.establishContext(PcscSCope.user);
      this.ctx = ctx;
    } catch (error) {
      throw transformPCSCError(error.toString());
    }
  }

  /// Get list of all readers
  ///
  /// * Throws if start() was not called before
  Future<List<String>> getReaders() async {
    checkInit();
    try {
      return await Pcsc.listReaders(ctx!);
    } catch (error) {
      throw transformPCSCError(error.toString());
    }
  }

  /// Connect a card inserted in a specific reader
  ///
  /// * Throws if start() was not called before
  Future<void> connectCard(String reader) async {
    checkInit();

    try {
      card = await Pcsc.cardConnect(
        ctx!,
        reader,
        PcscShare.shared,
        PcscProtocol.any,
      );
    } catch (error) {
      throw transformPCSCError(error.toString());
    }
  }

  /// Reads data from specified File on card
  ///
  /// * Throws if start() was not called before
  /// * Throws if connectCard() was not called
  @override
  Future<Map<CelikTag, String>> readData(CelikFile file) {
    checkInit();
    checkCard();

    return super.readData(file);
  }

  /// Reads image data.
  ///
  /// * Throws if start() was not called before
  /// * Throws if connectCard() was not called
  @override
  Future<String> readImageData() {
    checkInit();
    checkCard();

    return super.readImageData();
  }

  /// Automatically reads all data on card
  ///
  /// start() or end() does not have to be called!
  ///
  /// Returns always, no error throwing (Prints errors in kDebugMode)
  @override
  Future<CelikAPIAllData> readAllData() async {
    CelikAPIAllData data = CelikAPIAllData({}, "");
    try {
      await start();

      List<String> devices = await getReaders();
      if (devices.isEmpty) {
        throw "No devices found.";
      } else {
        await connectCard(devices[0]);
        data = await super.readAllData();
      }

      await end();
    } catch (e) {
      if (kDebugMode) {
        print("Error has occured:");
        print(e);
      }
    } finally {
      await end();
    }
    return data;
  }

  /// Disconnect a connected card
  ///
  /// * Throws if start() was not called before
  /// * Throws if connectCard() was not called
  Future<void> disconnectCard() async {
    checkInit();
    checkCard();

    try {
      await Pcsc.cardDisconnect(card!.hCard, PcscDisposition.resetCard);
      card = null;
    } catch (error) {
      throw transformPCSCError(error.toString());
    }
  }

  /// Called to terminate the connection to a card and PCSC Context
  ///
  /// Must be called after all the work is done to release context in memory
  Future<void> end() async {
    try {
      if (ctx != null) {
        if (card != null) {
          await Pcsc.cardDisconnect(card!.hCard, PcscDisposition.resetCard);
          card = null;
        }
        await Pcsc.releaseContext(ctx!);
        ctx = null;
      }
    } catch (error) {
      throw transformPCSCError(error.toString());
    }
  }
}


lib/src/celik_apdu_base.dart
import 'dart:math';
import 'dart:typed_data';

import './apdu.dart';
import './constants.dart';

/// Base class that defines APDU commands used by CelikAPI
///
/// One may override transmit method with a custom imiplementation
abstract class CelikAPDUBase {
  static const int defaultFileOffset = 4;
  dynamic selectedFileLength;
  CelikAPDUBase();

  Future<List<int>> transmit(List<int> bytes);

  Future<APDUResponse> init() =>
      transmit(APDUCommands.initAID(CelikConstants.lkAID))
          .then((b) => APDUResponse(b));

  Future<APDUResponse> selectFile(CelikFile file) =>
      transmit(APDUCommands.selectFile(CelikConstants.getFile(file))).then((b) {
        if (b.length > 2) selectedFileLength = b[3]; // Approximation
        return APDUResponse(b);
      });

  Future<APDUResponse> verifyPIN(List<int> pin) =>
      transmit(APDUCommands.verify(pin)).then((b) => APDUResponse(b));

  Future<bool> _setPreciseFileLength() async {
    APDUResponse res = APDUResponse(await transmit(
      APDUCommands.readByteData(2, 2),
    ));

    if (!res.isOk()) return false;

    selectedFileLength = Uint8List.fromList(res.data())
            .buffer
            .asByteData()
            .getUint16(0, Endian.little) +
        defaultFileOffset;

    return true;
  }

  Future<APDUResponse> readFile({int startOffset = defaultFileOffset}) async {
    if (selectedFileLength == null) return APDUResponse([103, 1]);

    List<int> data = [];
    List<int> statusCode = APDUStandard.successCode;

    if (!await _setPreciseFileLength()) {
      return APDUResponse([103, 1]);
    }

    int bytesOffset = startOffset;
    while (bytesOffset < selectedFileLength) {
      int bytesLength =
          min(selectedFileLength - bytesOffset, APDUStandard.maxReadBytes);

      APDUResponse res = APDUResponse(await transmit(
        APDUCommands.readByteData(bytesOffset, bytesLength),
      ));
      if (!res.isOk()) {
        statusCode = res.statusCode;
        break;
      }
      data.addAll(res.data());

      bytesOffset += bytesLength;
    }

    return APDUResponse([...data, ...statusCode]);
  }
}


lib/src/util.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';
import './constants.dart';

Map<int, List<int>> parseTLV(List<int> bytes) {
  Map<int, List<int>> out = {};

  int i = 0;
  while (i + 3 < bytes.length) {
    var indexStart = i;
    var valueLength = bytes[i + 2];
    var indexEnd = i + 4 + valueLength;

    if (bytes[i + 1] != 0x06 ||
        bytes[i + 3] != 0x00 ||
        indexEnd >= bytes.length) break;

    out[bytes[i]] = bytes.sublist(indexStart, indexEnd);
    i = indexEnd;
  }

  return out;
}

Map<CelikTag, String> transformData(List<int> data) {
  Map<CelikTag, String> transformed = {};
  ByteBuffer buf = Uint8List.fromList(data).buffer;

  int i = 0;
  while (i < data.length) {
    int tagId = buf.asByteData(i, 2).getUint16(0, Endian.little);

    int len = min(
      buf.asByteData(i + 2, 2).getUint16(0, Endian.little),
      data.length - i - 4, // Safe
    );

    String val = utf8.decode(buf.asUint8List(i + 4, len));
    CelikTag tag = CelikConstants.tagFromInt(tagId);
    if (tag != CelikTag.error) {
      transformed[tag] = val;
    } else {
      print("Cant find tag for $tagId");
    }

    i += len + 4;
  }

  return transformed;
}

/// Transforms errors thrown by `flutter_pcsc` package to readable string.
String transformPCSCError(String errorMessage) {
  final errorCode = errorMessage.substring(
    errorMessage.indexOf("Reason:") + "Reason: ".length,
  );
  return _transformPCSCErrorCode(errorCode);
}

String _transformPCSCErrorCode(String errorCode) {
  switch (errorCode) {
    case "SCARD_F_INTERNAL_ERROR":
      return 'Internal error.';
    case "SCARD_E_CANCELLED":
      return 'Cancelled.';
    case "SCARD_E_INVALID_HANDLE":
      return 'Invalid card handle.';
    case "SCARD_E_INVALID_PARAMETER":
      return 'Invalid parameter.';
    case "SCARD_E_INVALID_TARGET":
      return 'Invalid target.';
    case "SCARD_E_NO_MEMORY":
      return 'No memory.';
    case "SCARD_F_WAITED_TOO_LONG":
      return 'Waited for too long.';
    case "SCARD_E_INSUFFICIENT_BUFFER":
      return 'Insufficient buffer.';
    case "SCARD_E_UNKNOWN_READER":
      return 'Unknown reader device.';
    case "SCARD_E_TIMEOUT":
      return 'Timeout error!';
    case "SCARD_E_SHARING_VIOLATION":
      return 'Sharing violation!';
    case "SCARD_E_NO_SMARTCARD":
      return 'No smartcard present.';
    case "SCARD_E_UNKNOWN_CARD":
      return 'Unknown card.';
    case "SCARD_E_CANT_DISPOSE":
      return 'Error - Cannot dispose.';
    case "SCARD_E_PROTO_MISMATCH":
      return 'Error - Proto mismatch!';
    case "SCARD_E_NOT_READY":
      return 'Error - Not ready!';
    case "SCARD_E_INVALID_VALUE":
      return 'Invalid value!';
    case "SCARD_E_SYSTEM_CANCELLED":
      return 'Error - System cancelled.';
    case "SCARD_F_COMM_ERROR":
      return 'Communication error.';
    case "SCARD_F_UNKNOWN_ERROR":
      return 'Unknown error occurred.';
    case "SCARD_E_INVALID_ATR":
      return 'Invalid card ATR!';
    case "SCARD_E_NOT_TRANSACTED":
      return 'Error - Not transacted.';
    case "SCARD_E_READER_UNAVAILABLE":
      return 'Reader not available.';
    case "SCARD_P_SHUTDOWN":
      return 'Error - Shutdown.';
    case "SCARD_E_PCI_TOO_SMALL":
      return 'Error - PCI Too small.';
    case "SCARD_E_READER_UNSUPPORTED":
      return 'Reader not supported.';
    case "SCARD_E_DUPLICATE_READER":
      return 'Error - Duplicate reader.';
    case "SCARD_E_CARD_UNSUPPORTED":
      return 'Card is not supported.';
    case "SCARD_E_NO_SERVICE":
      return 'Error - No service.';
    case "SCARD_E_SERVICE_STOPPED":
      return 'Error - Service stopped.';
    case "SCARD_E_UNEXPECTED":
      return 'Unexpected error occurred.';
    case "SCARD_E_ICC_INSTALLATION":
      return 'Error - ICC Installation';
    case "SCARD_E_ICC_CREATEORDER":
      return 'Error - ICC Create Order';
    case "SCARD_E_UNSUPPORTED_FEATURE":
      return 'Error - Feature not supported.';
    case "SCARD_E_DIR_NOT_FOUND":
      return 'Error - Directory not found.';
    case "SCARD_E_FILE_NOT_FOUND":
      return 'Error - File not found.';
    case "SCARD_E_NO_DIR":
      return 'Error - No directory.';
    case "SCARD_E_NO_FILE":
      return 'Error - No file.';
    case "SCARD_E_NO_ACCESS":
      return 'Error - No access!';
    case "SCARD_E_WRITE_TOO_MANY":
      return 'Error - Too many writes!';
    case "SCARD_E_BAD_SEEK":
      return 'Error - Bad seek.';
    case "SCARD_E_INVALID_CHV":
      return 'Error - Invalid CHV!';
    case "SCARD_E_UNKNOWN_RES_MNG":
      return 'Error - Unknown res mng.';
    case "SCARD_E_NO_SUCH_CERTIFICATE":
      return 'Error - No such certificate!';
    case "SCARD_E_CERTIFICATE_UNAVAILABLE":
      return 'Error - Certificate not available!';
    case "SCARD_E_NO_READERS_AVAILABLE":
      return 'No readers available!';
    case "SCARD_E_COMM_DATA_LOST":
      return 'Error - Communication data lost!';
    case "SCARD_E_NO_KEY_CONTAINER":
      return 'Error - No key container.';
    case "SCARD_E_SERVER_TOO_BUSY":
      return 'Error - Server too busy!';
    case "SCARD_W_UNSUPPORTED_CARD":
      return 'Unsupported card.';
    case "SCARD_W_UNRESPONSIVE_CARD":
      return 'Non-responsive card.';
    case "SCARD_W_UNPOWERED_CARD":
      return 'Unpowered card.';
    case "SCARD_W_RESET_CARD":
      return 'Warning - Card was reset.';
    case "SCARD_W_REMOVED_CARD":
      return 'Warning - Card was removed.';
    case "SCARD_W_SECURITY_VIOLATION":
      return 'Card security violation!';
    case "SCARD_W_WRONG_CHV":
      return 'Wrong card CHV!';
    case "SCARD_W_CHV_BLOCKED":
      return 'Card CHV Blocked!';
    case "SCARD_W_EOF":
      return 'Card EOF!';
    case "SCARD_W_CANCELLED_BY_USER":
      return 'Cancelled by User.';
    case "SCARD_W_CARD_NOT_AUTHENTICATED":
      return 'Card not authenticated!';

    default:
      return 'Unknown error.';
  }
}
